### PR TITLE
Error overhaul to isolate User vs system errors

### PR DESCRIPTION
**Pull request by kumare3**
# TL;DR
This change enables every error in Propeller of Plugin to be classfied as a user/system/unknown error and expose it correctly to the higher level. This error is also preserved and can be retrieved from the CRD as well as from the Admin database and enables labeling metrics correctly
TODO: some inner function errors are not handled correctly yet, they may be incorrectly classified as system/unknown errors. (dynamic and branch)

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/lyft/flyte/issues/300

## Follow-up issue
https://github.com/lyft/flyte/issues/300

### CODE REVIEW

File: pkg/controller/nodes/dynamic/handler.go [hunk 1]
```diff
@@ -448,7 +450,7 @@ func (d dynamicNodeTaskNodeHandler) progressDynamicWorkflow(ctx context.Context,
 		}
 
 		return handler.DoTransition(handler.TransitionTypeEphemeral, handler.PhaseInfoRunning(nil)),
-			handler.DynamicNodeState{Phase: v1alpha1.DynamicNodePhaseFailing, Reason: state.Err.Error()},
+			handler.DynamicNodeState{Phase: v1alpha1.DynamicNodePhaseFailing, Reason: utils.ToString(state.Err)},
```
Comment (EngHabu) [inline pkg/controller/nodes/dynamic/handler.go:453]: I think you will have to preserve the Error here too (not serialize it) because that's what should be recorded as the error on the dynamic node?

File: pkg/controller/nodes/executor.go [hunk 2]
```diff
@@ -449,7 +451,7 @@ func (c *nodeExecutor) handleNode(ctx context.Context, dag executors.DAGStructur
 		}
 		nodeStatus.UpdatePhase(v1alpha1.NodePhaseFailed, v1.Now(), nodeStatus.GetMessage())
 		c.metrics.FailureDuration.Observe(ctx, nodeStatus.GetStartedAt().Time, nodeStatus.GetStoppedAt().Time)
-		return executors.NodeStatusFailed(fmt.Errorf(nodeStatus.GetMessage())), nil
+		return executors.NodeStatusFailed(utils.ParseExecutionError(nodeStatus.GetMessage())), nil
```
Comment (EngHabu) [inline pkg/controller/nodes/executor.go:454]: Not a big fan, can we just keep it a strongly typed exception all the way through?

Comment (kumare3) [reply pkg/controller/nodes/executor.go:454]: ohh i think I will have to implement the DeepCopy?

File: pkg/apis/flyteworkflow/v1alpha1/node_status.go [hunk 3]
```diff
@@ -180,6 +180,8 @@ type NodeStatus struct {
 
 	TaskNodeStatus    *TaskNodeStatus    `json:",omitempty"`
 	DynamicNodeStatus *DynamicNodeStatus `json:"dynamicNodeStatus,omitempty"`
+	// In case of Failing/Failed Phase, an execution error can be optionally associated with the Node
+	Err *ExecutionError `json:"error,omitempty"`
```
Comment (EngHabu) [inline pkg/apis/flyteworkflow/v1alpha1/node_status.go:184]: Nit: I would match the name in json and the field... 

```suggestion
	Error *ExecutionError `json:"error,omitempty"`
```

File: pkg/apis/flyteworkflow/v1alpha1/node_status.go [hunk 4]
```diff
@@ -396,6 +398,11 @@ func (in *NodeStatus) UpdatePhase(p NodePhase, occurredAt metav1.Time, reason st
 		n = metav1.Now()
 	}
 
+	if err != nil {
+		in.Err = new(ExecutionError)
```
Comment (EngHabu) [inline pkg/apis/flyteworkflow/v1alpha1/node_status.go:402]:
```suggestion
		in.Err = &ExecutionError { 
		                ExecutoinError: err,
		              }
```

File: pkg/apis/flyteworkflow/v1alpha1/node_status.go [hunk 5]
```diff
@@ -396,6 +398,11 @@ func (in *NodeStatus) UpdatePhase(p NodePhase, occurredAt metav1.Time, reason st
...
+		in.Err.ExecutionError = err
```
Comment (EngHabu) [inline pkg/apis/flyteworkflow/v1alpha1/node_status.go:403]:
```suggestion
```

File: pkg/apis/flyteworkflow/v1alpha1/workflow_status.go [hunk 6]
```diff
@@ -31,6 +32,9 @@ type WorkflowStatus struct {
 	// the retries the workflow will fail
 	FailedAttempts uint32 `json:"failedAttempts,omitempty"`
 
+	// Stores the Error during the Execution of the Workflow. It is optional and usually associated with Failing/Failed state only
+	Err *ExecutionError `json:"error,omitempty"`
```
Comment (EngHabu) [inline pkg/apis/flyteworkflow/v1alpha1/workflow_status.go:36]: nit:
```suggestion
	Error *ExecutionError `json:"error,omitempty"`
```

File: pkg/apis/flyteworkflow/v1alpha1/workflow_status.go [hunk 7]
```diff
@@ -55,13 +59,25 @@ func (in *WorkflowStatus) UpdatePhase(p WorkflowPhase, msg string) {
 		in.StartedAt = &n
 	}
 
+	if err != nil {
+		in.Err = new(ExecutionError)
```
Comment (EngHabu) [inline pkg/apis/flyteworkflow/v1alpha1/workflow_status.go:63]:
```suggestion
		in.Err = &ExecutionError {
		                  ExecutionError: err,
		              }
```

File: pkg/apis/flyteworkflow/v1alpha1/workflow_status.go [hunk 8]
```diff
@@ -55,13 +59,25 @@ func (in *WorkflowStatus) UpdatePhase(p WorkflowPhase, msg string) {
...
+		in.Err.ExecutionError = err
```
Comment (EngHabu) [inline pkg/apis/flyteworkflow/v1alpha1/workflow_status.go:64]:
```suggestion
```

File: pkg/controller/nodes/dynamic/handler.go [hunk 9]
```diff
@@ -98,7 +98,8 @@ func (d dynamicNodeTaskNodeHandler) handleParentNode(ctx context.Context, prevSt
 func (d dynamicNodeTaskNodeHandler) handleDynamicSubNodes(ctx context.Context, nCtx handler.NodeExecutionContext, prevState handler.DynamicNodeState) (handler.Transition, handler.DynamicNodeState, error) {
 	execContext, dynamicWF, nl, _, err := d.buildContextualDynamicWorkflow(ctx, nCtx)
 	if err != nil {
-		return handler.DoTransition(handler.TransitionTypeEphemeral, handler.PhaseInfoFailure(
+		// TODO @kumare classify system vs user errors
```
Comment (EngHabu) [inline pkg/controller/nodes/dynamic/handler.go:101]: Another PR?

File: pkg/controller/nodes/dynamic/handler.go [hunk 10]
```diff
@@ -441,42 +442,35 @@ func (d dynamicNodeTaskNodeHandler) progressDynamicWorkflow(ctx context.Context,
...
+	if state.HasFailed() || state.HasTimedOut() {
+		// When the subworkflow either fails or times-out we need to handle failing
 		if dynamicWorkflow.GetOnFailureNode() != nil {
 			// TODO Once we migrate to closure node we need to handle subworkflow using the subworkflow handler
 			logger.Errorf(ctx, "We do not support failure nodes in dynamic workflow today")
 		}
 
-		return handler.DoTransition(handler.TransitionTypeEphemeral, handler.PhaseInfoRunning(nil)),
-			handler.DynamicNodeState{Phase: v1alpha1.DynamicNodePhaseFailing, Reason: state.Err.Error()},
+		// As we do not support Failure Node, we can just return failure in this case
```
Comment (EngHabu) [inline pkg/controller/nodes/dynamic/handler.go:452]: There was a good reason why we added this extra step in the internal statemachine, I wish I documented that :) let me try to remember, this node requires essentially two rounds to abort, since nodes inside it will first transition to Failing in one round, then their aborts will be called in the next? or maybe that was just an artifact for not having recursive abort? I would rather keep this change in a separate PR though. It seems risky if we get it wrong... and I would like this PR to go in ASAP.

File: pkg/controller/nodes/dynamic/handler.go [hunk 11]
```diff
@@ -441,42 +442,35 @@ func (d dynamicNodeTaskNodeHandler) progressDynamicWorkflow(ctx context.Context,
...
+		return handler.DoTransition(handler.TransitionTypeEphemeral, handler.PhaseInfoFailureErr(state.Err, nil)),
+			handler.DynamicNodeState{Phase: v1alpha1.DynamicNodePhaseFailing, Reason: "Dynamic Workflow execution Failed"},
```
Comment (EngHabu) [inline pkg/controller/nodes/dynamic/handler.go:454]: Do we need to add ExecutionError in the DynamcNodeState to be able to propagate it?

File: pkg/controller/nodes/subworkflow/launchplan.go [hunk 12]
```diff
@@ -98,15 +98,15 @@ func (l *launchPlanHandler) CheckLaunchPlanStatus(ctx context.Context, nCtx hand
...
+		return handler.DoTransition(handler.TransitionTypeEphemeral, handler.PhaseInfoFailure(core.ExecutionError_USER, errors.RemoteChildWorkflowExecutionFailed, err.Error(), &handler.ExecutionInfo{
 			WorkflowNodeInfo: &handler.WorkflowNodeInfo{LaunchedWorkflowID: childID},
 		})), nil
 	case core.WorkflowExecution_FAILED:
 		errMsg := fmt.Sprintf("launchplan execution failed without explicit error")
 		if wfStatusClosure.GetError() != nil {
 			errMsg = fmt.Sprintf(" errorCode[%s]: %s", wfStatusClosure.GetError().Code, wfStatusClosure.GetError().Message)
 		}
-		return handler.DoTransition(handler.TransitionTypeEphemeral, handler.PhaseInfoFailure(errors.RemoteChildWorkflowExecutionFailed, errMsg, &handler.ExecutionInfo{
+		return handler.DoTransition(handler.TransitionTypeEphemeral, handler.PhaseInfoFailure(core.ExecutionError_USER, errors.RemoteChildWorkflowExecutionFailed, errMsg, &handler.ExecutionInfo{
```
Comment (EngHabu) [inline pkg/controller/nodes/subworkflow/launchplan.go:109]: The execution closure error has kind. should we use the kind from there? If it failed due to a system error, should this execution also fail with a system error?

File: pkg/apis/flyteworkflow/v1alpha1/node_status.go (hunk 3 above)
Comment (kumare3) [reply pkg/apis/flyteworkflow/v1alpha1/node_status.go:184]: I guess that should work, lower case `error` is a reserved keyword

File: pkg/controller/nodes/dynamic/handler.go (hunk 1 above)
Comment (kumare3) [reply pkg/controller/nodes/dynamic/handler.go]: removed

File: pkg/controller/nodes/executor.go (hunk 2 above)
Comment (kumare3) [reply pkg/controller/nodes/executor.go]: done

File: pkg/controller/nodes/subworkflow/launchplan.go (hunk 12 above)
Comment (kumare3) [reply pkg/controller/nodes/subworkflow/launchplan.go:109]: yes we should

File: pkg/controller/nodes/dynamic/handler.go (hunk 11 above)
Comment (kumare3) [reply pkg/controller/nodes/dynamic/handler.go:454]: I thought about. I dont think so, as the handler is already returning an error

File: pkg/controller/nodes/dynamic/handler.go (hunk 10 above)
Comment (kumare3) [reply pkg/controller/nodes/dynamic/handler.go:452]: ok

File: pkg/controller/nodes/dynamic/handler.go (hunk 9 above)
Comment (kumare3) [reply pkg/controller/nodes/dynamic/handler.go:101]: yes

File: pkg/controller/nodes/dynamic/handler.go (hunk 10 above)
Comment (kumare3) [reply pkg/controller/nodes/dynamic/handler.go:452]: the only thing is, if we want to keep the state running, then i will need to add execution error to dynamic node state

File: pkg/controller/nodes/dynamic/handler.go (hunk 9 above)
Comment (kumare3) [reply pkg/controller/nodes/dynamic/handler.go:101]: yes, added as a tODO to this PR

File: pkg/controller/executors/node_lookup.go [hunk 13]
```diff
@@ -14,24 +14,6 @@ type NodeLookup interface {
 	GetNodeExecutionStatus(ctx context.Context, id v1alpha1.NodeID) v1alpha1.ExecutableNodeStatus
 }
 
-// Implements a de-generate case of NodeLookup, where only one Node is always looked up
```
Comment (wild-endeavor) [inline pkg/controller/executors/node_lookup.go:17]: Why are we removing this?

Comment (kumare3) [reply pkg/controller/executors/node_lookup.go:17]: we dont use it, we can add it when we need it

File: pkg/controller/nodes/dynamic/handler.go (hunk 10 above)
Comment (EngHabu) [reply pkg/controller/nodes/dynamic/handler.go:452]: Agreed... let's do that... in the Dynamic-PR

File: pkg/controller/nodes/subworkflow/launchplan.go [hunk 14]
```diff
@@ -44,7 +44,7 @@ func (l *launchPlanHandler) StartLaunchPlan(ctx context.Context, nCtx handler.No
 		if launchplan.IsAlreadyExists(err) {
 			logger.Infof(ctx, "Execution already exists [%s].", childID.Name)
 		} else if launchplan.IsUserError(err) {
-			return handler.DoTransition(handler.TransitionTypeEphemeral, handler.PhaseInfoFailure(errors.RuntimeExecutionError, "failed to create unique ID", &handler.ExecutionInfo{
+			return handler.DoTransition(handler.TransitionTypeEphemeral, handler.PhaseInfoFailure(core.ExecutionError_USER, errors.RuntimeExecutionError, "failed to create unique ID", &handler.ExecutionInfo{
```
Comment (EngHabu) [inline pkg/controller/nodes/subworkflow/launchplan.go:47]: the error description seems off. no? we should copy the string from `err.String()` probably?

File: pkg/controller/nodes/subworkflow/launchplan.go [hunk 15]
```diff
@@ -69,15 +69,15 @@ func (l *launchPlanHandler) CheckLaunchPlanStatus(ctx context.Context, nCtx hand
...
+		return handler.DoTransition(handler.TransitionTypeEphemeral, handler.PhaseInfoFailure(core.ExecutionError_SYSTEM, errors.RuntimeExecutionError, "failed to create unique ID", nil)), nil
 	}
 
 	wfStatusClosure, err := l.launchPlan.GetStatus(ctx, childID)
 	if err != nil {
 		if launchplan.IsNotFound(err) { // NotFound
 			errorCode, _ := errors.GetErrorCode(err)
 			err = errors.Wrapf(errorCode, nCtx.NodeID(), err, "launch-plan not found")
-			return handler.DoTransition(handler.TransitionTypeEphemeral, handler.PhaseInfoFailure(errorCode, err.Error(), &handler.ExecutionInfo{
+			return handler.DoTransition(handler.TransitionTypeEphemeral, handler.PhaseInfoFailure(core.ExecutionError_USER, errorCode, err.Error(), &handler.ExecutionInfo{
```
Comment (EngHabu) [inline pkg/controller/nodes/subworkflow/launchplan.go:80]: hmm is this a user error? if we come here we are saying the system guaranteed we launched something, right? it can't just disappear by any user action... not that we ever hit this error

File: pkg/controller/nodes/subworkflow/launchplan.go [hunk 16]
```diff
@@ -98,15 +98,15 @@ func (l *launchPlanHandler) CheckLaunchPlanStatus(ctx context.Context, nCtx hand
 	case core.WorkflowExecution_ABORTED:
 		wErr = fmt.Errorf("launchplan execution aborted")
 		err = errors.Wrapf(errors.RemoteChildWorkflowExecutionFailed, nCtx.NodeID(), wErr, "launchplan [%s] failed", childID.Name)
```
Comment (EngHabu) [inline pkg/controller/nodes/subworkflow/launchplan.go:100]: Shouldn't this message say "aborted" ?

File: pkg/controller/nodes/subworkflow/launchplan.go (hunk 15 above)
Comment (kumare3) [reply pkg/controller/nodes/subworkflow/launchplan.go:80]: hmm but then this should not happen. It could be a system error where we have a corruption

File: pkg/controller/nodes/subworkflow/launchplan.go (hunk 14 above)
Comment (kumare3) [reply pkg/controller/nodes/subworkflow/launchplan.go:47]: done
